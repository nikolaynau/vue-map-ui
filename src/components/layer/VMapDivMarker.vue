<script setup lang="ts">
import { getCurrentInstance, useSlots } from 'vue';
import type {
  CrossOrigin,
  DivIcon,
  LatLngExpression,
  Marker,
  MarkerOptions,
  PointExpression
} from 'leaflet';
import { useTemplateRef } from '../../composables/internal/useTemplateRef';
import { splitProps } from '../../utils/props';
import VMapMarker from './VMapMarker.vue';
import VMapDivIcon from './VMapDivIcon.vue';

export interface Props extends MarkerOptions {
  latlng: LatLngExpression | null;
  iconHtml?: string | HTMLElement | false;
  iconBgPos?: PointExpression;
  iconRenderMode?: 'html' | 'node' | 'portal' | 'none';
  iconRootClass?: any;
  iconUrl?: string;
  iconRetinaUrl?: string;
  iconSize?: PointExpression;
  iconAnchor?: PointExpression;
  iconShadowUrl?: string;
  iconShadowRetinaUrl?: string;
  iconShadowSize?: PointExpression;
  iconShadowAnchor?: PointExpression;
  iconKnownClasses?: string[];
  iconClass?: any;
  iconClassName?: any;
  iconCrossOrigin?: CrossOrigin | boolean;
  iconPopupAnchor?: PointExpression;
  iconTooltipAnchor?: PointExpression;
  iconPane?: string;
  iconAttribution?: string;
}

const props = defineProps<Props>();

const instance = getCurrentInstance()!;
const splitted = splitProps(instance, props, 'icon', ['latlng'], true, true, [
  'iconRetinaUrl',
  'iconSize',
  'iconAnchor'
]);

const slots = useSlots() as { default: unknown };

const { templateRef: markerRef, value: marker } = useTemplateRef<
  InstanceType<typeof VMapMarker>,
  Marker | null
>(obj => obj.marker);

const { templateRef: iconRef, value: icon } = useTemplateRef<
  InstanceType<typeof VMapDivIcon>,
  DivIcon | null
>(obj => obj.icon);

defineExpose({
  marker,
  icon
});
</script>

<template>
  <VMapMarker ref="markerRef" :latlng="latlng" v-bind="splitted.rest">
    <VMapDivIcon ref="iconRef" v-bind="splitted.matching">
      <template v-if="slots.default" #default>
        <slot></slot>
      </template>
    </VMapDivIcon>
  </VMapMarker>
</template>
