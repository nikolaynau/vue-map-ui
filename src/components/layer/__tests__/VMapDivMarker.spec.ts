import { describe, it, expect } from 'vitest';
import { ref, reactive, h, defineComponent, onMounted, nextTick } from 'vue';
import { DivIcon, Marker } from 'leaflet';
import { mount } from '../../../../.test';
import VMapDivMarker from '../VMapDivMarker.vue';

describe('VMapDivMarker', () => {
  it('should be expose instances', () => {
    expect.assertions(2);

    const Component = defineComponent({
      setup() {
        const marker = ref<InstanceType<typeof VMapDivMarker> | null>(null);

        onMounted(() => {
          expect(marker.value?.icon).toBeInstanceOf(DivIcon);
          expect(marker.value?.marker).toBeInstanceOf(Marker);
        });

        return () => h(VMapDivMarker, { latlng: [0, 0], ref: marker });
      }
    });

    mount(Component);
  });

  it('should be reactive attrs', async () => {
    const props = reactive<any>({
      latlng: [0, 0],
      iconHtml: '<i class="icon">Foo</i>'
    });

    const marker = ref<InstanceType<typeof VMapDivMarker> | null>(null);

    const Component = defineComponent({
      setup() {
        return () => h(VMapDivMarker, { ref: marker, ...props });
      }
    });

    mount(Component);

    await nextTick();

    expect(marker.value?.marker?.getLatLng()).toEqual({ lat: 0, lng: 0 });
    expect(marker.value?.icon?.options.html).toMatchSnapshot('init html');

    props.latlng = [1, 2];
    props.iconHtml = '<i class="icon">Bar</i>';
    await nextTick();

    expect(marker.value?.marker?.getLatLng()).toEqual({ lat: 1, lng: 2 });
    expect(marker.value?.icon?.options.html).toMatchSnapshot('change html');
  });

  it('should be render default slot', async () => {
    const marker = ref<InstanceType<typeof VMapDivMarker> | null>(null);

    const Component = defineComponent({
      setup() {
        return () =>
          h(VMapDivMarker, { ref: marker, latlng: [0, 0] }, () =>
            h('i', { class: 'icon' })
          );
      }
    });

    mount(Component);

    await nextTick();
    expect(marker.value?.icon?.options.html).toMatchSnapshot();
  });
});
