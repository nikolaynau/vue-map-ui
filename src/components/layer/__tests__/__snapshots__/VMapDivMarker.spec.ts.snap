// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VMapDivMarker > should be reactive attrs > change html 1`] = `"<i class="icon">Bar</i>"`;

exports[`VMapDivMarker > should be reactive attrs > init html 1`] = `"<i class="icon">Foo</i>"`;

exports[`VMapDivMarker > should be render default slot 1`] = `"<i class="icon"></i>"`;
